library/core/src/option.rs

```diff
@@ -90,6 +90,321 @@
 //! from `Some::<T>(_)` to `T` (but transmuting `None::<T>` to `T`
 //! is undefined behaviour).
 //!
+//! # Method overview
+//!
+//! In addition to working with pattern matching, [`Option`] provides a wide
+//! variety of different methods.
+//!
+//! ## Querying the variant
+//!
+//! The [`is_some`] and [`is_none`] methods return [`true`] if the [`Option`]
+//! is [`Some`] or [`None`], respectively.
+//!
+//! [`is_some`]: Option::is_some
+//! [`is_none`]: Option::is_none
+//!
+//! ## Adapters for working with references
+//!
+//! * [`as_ref`] converts from `&Option<T>` to `Option<&T>`
+//! * [`as_mut`] converts from `&mut Option<T>` to `Option<&mut T>`
+//! * [`as_deref`] converts from `&Option<T>` to `Option<&T::Target>`
+//! * [`as_deref_mut`] converts from `&mut Option<T>` to `Option<&mut T::Target>`
+//! * [`as_pin_ref`] converts from [`&Pin`]`<Option<T>>` to `Option<`[`Pin`]`<&T>>`
+//! * [`as_pin_mut`] converts from [`&mut Pin`]`<Option<T>>` to `Option<`[`Pin`]`<&mut T>>`
+//!
+//! [`&mut Pin`]: crate::pin::Pin
+//! [`&Pin`]: crate::pin::Pin
+//! [`as_deref`]: Option::as_deref
+//! [`as_deref_mut`]: Option::as_deref_mut
+//! [`as_mut`]: Option::as_mut
+//! [`as_pin_ref`]: Option::as_pin_ref
+//! [`as_pin_mut`]: Option::as_pin_mut
+//! [`as_ref`]: Option::as_ref
+//! [`Pin`]: crate::pin::Pin
+//!
+//! ## Extracting the contained value
+//!
+//! These methods extract the contained value in an [`Option`] when it is
+//! the [`Some`] variant. If the [`Option`] is [`None`]:
+//!
+//! * [`expect`] panics with a provided custom message
+//! * [`unwrap`] panics with a generic message
+//! * [`unwrap_or`] returns the provided default value
+//! * [`unwrap_or_default`] returns the default value of the type `T`
+//!   (which must implement the [`Default`] trait)
+//! * [`unwrap_or_else`] returns the result of evaluating the provided
+//!   function
+//!
+//! [`Default`]: crate::default::Default
+//! [`expect`]: Option::expect
+//! [`unwrap`]: Option::unwrap
+//! [`unwrap_or`]: Option::unwrap_or
+//! [`unwrap_or_default`]: Option::unwrap_or_default
+//! [`unwrap_or_else`]: Option::unwrap_or_else
+//!
+//! ## Transforming contained values
+//!
+//! These transformations are from [`Option`] to [`Result`].
+//!
+//! * [`ok_or`] transforms [`Some(v)`] to [`Ok(v)`], and [`None`] to
+//!   [`Err(err)`] using the provided default `err` value
+//! * [`ok_or_else`] transforms [`Some(v)`] to [`Ok(v)`], and [`None`] to
+//!   a value of [`Err`] using the provided function
+//! * [`transpose`] transposes an [`Option`] of a [`Result`] into a
+//!   [`Result`] of an [`Option`]
+//!
+//! These transformations are on [`Some`] values.
```

Comment:
This seems like a misleading heading; some of these functions (`map_or` and `map_or_else`) affect `None` values as well.

Also:

```suggestion
//! These methods transform the [`Some`] variant:
```

Reply:
I think of the `map` family of functions as operating primarily on the content of `Some` values, with `map_or` and `map_or_else` providing fallback values in the case of `None`. Alternatively, we could categorize `map` on its own, because it maps `Some(x)` to `Some(y)`, while `map_or` and `map_or_else` can produce arbitrary types. What do you think?

Reply:
I think the latter makes sense; `map_or` and `map_or_else` are effectively a one-function match statement, specifying how to handle both Some and None (or for a Result, how to handle both Ok and Err).

Reply:
It's somewhat at odds with the current naming of those methods; maybe `unwrap_map_or` and `unwrap_map_or_else` would be more accurate, but also too verbose. `map` transforms "inside" a `Some`, while the others don't necessarily return an `Option` at all, so having similar names seems weird.